packages/caravan-fees/src/utils.ts

```diff
@@ -0,0 +1,466 @@
+      const p2wpkhAddress = payments.p2wpkh({
+        output: script,
+        network: bitcoinjsNetwork,
+      }).address;
+      if (p2wpkhAddress) return p2wpkhAddress;
+    }
+
+    // Check for P2WSH
+    if (script.length === 34 && script[0] === 0x00 && script[1] === 0x20) {
+      const p2wshAddress = payments.p2wsh({
+        output: script,
+        network: bitcoinjsNetwork,
+      }).address;
+      if (p2wshAddress) return p2wshAddress;
+    }
+
+    // Check for P2SH (23 bytes, starting with OP_HASH160 (0xa9))
+    if (script.length === 23 && script[0] === 0xa9 && script[22] === 0x87) {
+      return (
+        payments.p2sh({ output: script, network: bitcoinjsNetwork }).address ||
+        ""
+      );
+    }
+
+    // Check for P2TR (Taproot, 34 bytes, starting with OP_1 (0x51))
+    if (script.length === 34 && script[0] === 0x51) {
+      const p2trAddress = payments.p2tr({
+        output: script,
+        network: bitcoinjsNetwork,
+      }).address;
+      if (p2trAddress) return p2trAddress;
+    }
+
+    // If we couldn't derive an address, return an error message
+    return "Unable to derive address";
+  } catch (e) {
+    console.error("Error deriving address:", e);
+    return "Unable to derive address";
+  }
+}
+
+/**
+ * Estimates the virtual size (vsize) of a transaction.
+ *
+ * This function calculates the estimated vsize for different address types,
+ * including P2SH, P2SH_P2WSH, P2WSH, and P2WPKH. The vsize is crucial for
+ * fee estimation in Bitcoin transactions, especially for SegWit transactions
+ * where the witness data is discounted.
+ *
+ * Calculation Method:
+ * 1. For non-SegWit (P2SH): vsize = transaction size
+ * 2. For SegWit (P2SH_P2WSH, P2WSH, P2WPKH):
+ *    vsize = (transaction weight) / 4, rounded up
+ *    where transaction weight = (base size * 3) + total size
+ *
+ * References:
+ * - BIP141 (SegWit): https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki
+ * - Bitcoin Core weight calculation: https://github.com/bitcoin/bitcoin/blob/master/src/consensus/validation.h
+ *
+ * @param config - Configuration object for the transaction
+ * @param config.addressType - The type of address used (P2SH, P2SH_P2WSH, P2WSH, P2WPKH)
+ * @param config.numInputs - Number of inputs in the transaction
+ * @param config.numOutputs - Number of outputs in the transaction
+ * @param config.m - Number of required signatures (for multisig)
+ * @param config.n - Total number of possible signers (for multisig)
+ *
+ * @returns The estimated virtual size (vsize) of the transaction in vbytes
+ *
+ * @throws Will throw an error if the address type is unsupported
+ */
+export function estimateTransactionVsize(
+  config: {
+    addressType?: ScriptType;
+    numInputs?: number;
+    numOutputs?: number;
+    m?: number;
+    n?: number;
+  } = {},
+): number {
+  const {
+    addressType = SCRIPT_TYPES.P2SH,
+    numInputs = 1,
+    numOutputs = 1,
+    m = 1,
+    n = 2,
+  } = config;
```

Comment:
This could all be done in the function signature with destructuring.

Reply:
I refactor that , sorry that's just a bad coding habit I have `:)

Reply:
changes [done]([Fix-Utils : estimateTransactionVsize](https://github.com/caravan-bitcoin/caravan/pull/114/commits/4f6b06773d1ea78b50f7858e054d6a7f8d10171e))